### DISCUSSION TITLE
Tech

### DISCUSSION THREAD

**Discussion by Roshanjossey** [jina-ai/dashboard#127] [General]
A living document on where we're headed in terms of tech of console and things we want to achieve along the way.

### Problems to solve in the future.

1. Localisation: We should think about making console more accessible to people who aren't fluent with English. 
2. Widget architecture: We might move in a direction where console is highly customisable with widgets users can add and remove.
3. Performance optimisations: We could use `useMemo`, `useCallback` etc to optimize by avoiding re-renders and re-declarations. But let's profile before optimizing
4. [Reselect](https://github.com/reduxjs/reselect): If our state selection logic becomes more complex, we might have to find ways to memoize selectors.

### Managing tech debt.

For moving fast, we have borrowed some time from our future selves. Let's be mindful of what we owe and try to payback and reduce the debt.

This section is to keep track of what we want to improve in terms of tech and reliability of our code. One guideline we could keep in our mind is to improve the parts we touch. For each component/ function/ file we use or update, let's be mindful of the following.

1. Test coverage: Currently, we're at 12.12% in terms of unit test coverage. We should target a much healthier coverage.
2. Typing: There are places where we're using `any` or could use stricter rules. 
3. Styling: With the changes for theming, we're going with css-in-js instead of sass/scss. Let's try to refactor styles and use more component styles

**Done**

- [x]  state management: We've made a decision to move to redux from flux. Let's try to move more parts of our state management code to redux.

**Comment by DavidSanwald** [comment]
I think those are valid points. Some thoughts on the last for points you mentioned:
1. Good point. I think once the project got started with component tests, it will get easier to add further tests.  I would consider choosing a small number of components, prioritizing those that are reused and adapted to multiple places in the codebase, and assign some story points.
2. I think the typing has come a long way. It's already much better than it has been.  Kudos to the team! 👍  I think the hardest thing is interfacing with code that uses a lot of `any` without letting it spread into the code you're currently writing. In the end, there will always be time pressure. 
I think it would be very helpful to consider using unknown instead of any, when possible, and make more frequent use of custom type guards.

Will address the remaining points later.

**Comment by Roshanjossey** [comment]
We've made some good progress on test coverage. Now, we're sitting at 33%. https://app.codecov.io/gh/jina-ai/dashboard
Looking at the breakdown, I think that our concentration of tests for state management, helpers and services is good.

**Comment by DavidSanwald** [comment]
Obviously, there's a team of talented and hard-working people on the front end.
Kudos to the team the dashboard has come a long way.

Some suggestion:
If there are certain parts of the pipeline that fail but failing doesn't mean something won't be merged the team should discuss that.
I think that's problematic. If those steps in the pipeline matter, code that causes those steps to fail should not be merged or even shouldn't be possible to merge. If those steps don't matter, they should be removed from the pipeline.

Some pipeline that runs, maybe says something but not really or only depending on certain circumstances can cause problems long term.
